avoid module-level initLogging for non-interactive CLI commands

bertsky opened the pull request:
Fix #589 (instead of #591)

This seems to work. Not so sure about the placement of `initLogging()` at the module level of `ocrd.cli.*` though...

bertsky commented:
BTW, a simple way to test this is setting `logger_root` to `level=DEBUG` in `ocrd_logging.conf` (in `/etc` if in Docker or `~` native) – without this PR, you see DLL debug messages for `shapely.geos` in `ocrd --help` or `ocrd --version` (or similar). I tried to identify all CLI contexts which must not normally produce logging output in the patch (but might have missed some).

kba reviewed (approved):
Looks good. I'm not sure how intuitive I find it that importing `ocrd_utils.logging` effectively *disables* logging or that we now must do `initLogging` before doing anything else but at least that is well-defined behavior.

bertsky commented on ocrd_utils/ocrd_utils/logging.py:
```diff
@@ -57,7 +57,7 @@ def getLevelName(lvl):
     lvl = _ocrdLevel2pythonLevel.get(lvl, lvl)
     return logging.getLevelName(lvl)
 
-def setOverrideLogLevel(lvl, silent=False):
```
Sorry, I thought this would not be needed anymore. I'll revert that part.

bertsky replied:
Done. (I have also added more comments.)

bertsky commented:
> I'm not sure how intuitive I find it that importing `ocrd_utils.logging` effectively _disables_ logging or that we now must do `initLogging` before doing anything else but at least that is well-defined behavior.

Yes, it's not intuitive. The only other option I can imagine to get rid of logging in contexts like `--version`, `--help`, `--dump-json`, `ocrd-tool ...`, `bashlib ...` would be not changing ocrd_logging, but _not importing it at the module level_ anymore (but only when in a context where we really want it).

Also, we would still have to change logging setup/teardown for regression tests for this PR...

bertsky commented:
> Yes, it's not intuitive. The only other option I can imagine to get rid of logging in contexts like `--version`, `--help`, `--dump-json`, `ocrd-tool ...`, `bashlib ...` would be not changing ocrd_logging, but _not importing it at the module level_ anymore (but only when in a context where we really want it).

But IIUC that would still not protect us against bad apple modules/libraries (like Matplotlib, Tensorflow...) which setup their logging in a way that they would interfere even without `basicConfig` from our side.

bertsky commented:
> Also, we would still have to change logging setup/teardown for regression tests for this PR...

I must confess I am perplexed about the 2 CI failures.
> `08:56:50.129 INFO ocrd.process.profile - [232 chars]\'\n`

Where could this possibly be originating from? `ocrd.processor.helpers.run_processor` only uses that logger once, with a completely different message content (`'Executing processor'...`)

kba commented:
> I must confess I am perplexed about the 2 CI failures.
> 
> > `08:56:50.129 INFO ocrd.process.profile - [232 chars]\'\n`
> 
> Where could this possibly be originating from? `ocrd.processor.helpers.run_processor` only uses that logger once, with a completely different message content (`'Executing processor'...`)

The log messages themselves make sense, you're quoting the pytest abbreviation, the full output is

```
{"baz": "two"}
11:12:41.779 INFO ocrd.process.profile - Executing processor 'ocrd-test' took 0.000027s [--input-file-grp='OCR-D-IMG' --output-file-grp='OUTPUT' --parameter='{"baz": "two"}']
11:12:41.779 INFO ocrd.workspace - Saving mets '/tmp/ocrd-test-tempdir-boa5wmy4/mets.xml'
```

Now that initLogging needs to be explicitly called, these log messages interfere with the output of the processor.

I cannot, however, reproduce this locally. Can you?

bertsky commented:
> The log messages themselves make sense, you're quoting the pytest abbreviation, the full output is

Oh, where do you get these from? Even the [download link](https://circleci.com/api/v1.1/project/github/OCR-D/core/1908/output/103/0?file=true&allocation-id=5f55f58b46f831119cf20ba5-0-build%2FF6B90B4) does not show this...

> Now that initLogging needs to be explicitly called, these log messages interfere with the output of the processor.

That's what I don't understand. Why didn't we get output from `ocrd.process.profile` when we used to `initLogging` unconditionally? It should never have worked ...

This test simulates a true CLI call, included standard logging (however implemented). In our unit tests, we can either test for non-logging output (where we should _disable_ logging) or for log output (with specific loggers _enabled_).

> I cannot, however, reproduce this locally. Can you?

No, not locally. All checks succeed.

bertsky commented:
It looks like this is related to the differences between unittest's and click's runner. I have rebased and added a change to test that hypothesis. Thumbs crossed for CI... (Yes, that still does not give us conscious control over what kind of output we want in each test, but let's first get green lights on.)

bertsky commented:
That _did_ work! Do you remember the reason for dealing with two different runners? (I fail to see the pattern.) Should we revisit tests as to whether or not they should or should not contain log output next, or are you satisfied with the current state for now?

> > Yes, it's not intuitive. The only other option I can imagine to get rid of logging in contexts like `--version`, `--help`, `--dump-json`, `ocrd-tool ...`, `bashlib ...` would be not changing ocrd_logging, but _not importing it at the module level_ anymore (but only when in a context where we really want it).
> 
> But IIUC that would still not protect us against bad apple modules/libraries (like Matplotlib, Tensorflow...) which setup their logging in a way that they would interfere even without `basicConfig` from our side.

I am not so enthusiastic about this PR myself, but what else could we do?

kba commented:
> I am not so enthusiastic about this PR myself, but what else could we do?

Ah, cannot win prizes for elegance every time :) It's a good solution.

> Do you remember the reason for dealing with two different runners?

Output management, c.f. tests/base.py. pytest captures STDERR/STDOUT and log messages (to show in case of failure). In my experience, the click CliRunner works for simple tasks but for everything that requires STDOUT/STDERR (especially: separate), it's easier to run `cli.main` and catch SystemExit.

bertsky commented:
> Output management, c.f. tests/base.py. pytest captures STDERR/STDOUT and log messages (to show in case of failure). In my experience, the click CliRunner works for simple tasks but for everything that requires STDOUT/STDERR (especially: separate), it's easier to run `cli.main` and catch SystemExit.

Understood. Then the last change is indeed warranted. (And the only place where log messages are checked via CLI is AFAICS  `TestDecorators.test_loglevel_invalid` – using CliRunner – and `TestLogging.testProcessorProfiling` – using `ocrd.run_processor` directly but also custom stream handler.)
